script.tvmaze.integration/resources/lib/tvmi.py

```diff
@@ -0,0 +1,617 @@
+        self.TVMAZE = tvmaze.API( user=self.SETTINGS['tvmaze_user'], apikey=self.SETTINGS['tvmaze_apikey'] )
+        self.TVMCACHEFILE = os.path.join( self.SETTINGS['ADDONDATAPATH'], 'tvm_followed_cache.json' )
+        self.DIALOG = xbmcgui.Dialog()
+        self.KODIMONITOR = xbmc.Monitor()
+
+
+    def _option_follow_shows( self ):
+        showlist = self._build_show_list()
+        ret = self._select_shows_dialog( self.SETTINGS['ADDONLANGUAGE']( 32202 ), showlist )
+        if not ret:
+            return
+        else:
+            xbmc.executebuiltin( 'ActivateWindow(busydialognocancel)' )
+            self._add_shows( ret, showlist )
+            xbmc.executebuiltin( 'Dialog.Close(busydialognocancel)' )
+
+
+    def _option_unfollow_shows( self ):
+        xbmc.executebuiltin( 'ActivateWindow(busydialognocancel)' )
+        followedlist, followedmap = self._build_tvmaze_list()
+        xbmc.executebuiltin( 'Dialog.Close(busydialognocancel)' )
+        ret = self._select_shows_dialog( self.SETTINGS['ADDONLANGUAGE']( 32203 ), followedlist )
+        if not ret:
+            return
+        else:
+            xbmc.executebuiltin( 'ActivateWindow(busydialognocancel)' )
+            self._unfollow_shows( ret, followedlist, followedmap )
+            xbmc.executebuiltin( 'Dialog.Close(busydialognocancel)' )
+
+
+    def _option_tag_shows( self ):
+        xbmc.executebuiltin( 'ActivateWindow(busydialognocancel)' )
+        taglist, tagmap = _build_tag_list( self.TVMAZE, self.LW )
+        xbmc.executebuiltin( 'Dialog.Close(busydialognocancel)' )
+        ret = self.DIALOG.select( self.SETTINGS['ADDONLANGUAGE']( 32204 ), taglist )
+        if ret == -1:
+            return
+        tagid = tagmap[taglist[ret]]
+        showlist = self._build_show_list()
+        ret = self._select_shows_dialog( self.SETTINGS['ADDONLANGUAGE']( 32205 ), showlist )
+        if not ret:
+            return
+        else:
+            xbmc.executebuiltin( 'ActivateWindow(busydialognocancel)' )
+            self._add_shows( ret, showlist, tagid=tagid )
+            xbmc.executebuiltin( 'Dialog.Close(busydialognocancel)' )
+
+
+    def _option_untag_shows( self ):
+        xbmc.executebuiltin( 'ActivateWindow(busydialognocancel)' )
+        taglist, tagmap = _build_tag_list( self.TVMAZE, self.LW )
+        xbmc.executebuiltin( 'Dialog.Close(busydialognocancel)' )
+        ret = self.DIALOG.select( self.SETTINGS['ADDONLANGUAGE']( 32204 ), taglist )
+        if ret == -1:
+            return
+        tagid = tagmap[taglist[ret]]
+        xbmc.executebuiltin( 'ActivateWindow(busydialognocancel)' )
+        taggedlist, taggedmap = self._build_tvmaze_list( tagid=tagid )
+        xbmc.executebuiltin( 'Dialog.Close(busydialognocancel)' )
+        ret = self._select_shows_dialog( self.SETTINGS['ADDONLANGUAGE']( 32206 ), taggedlist )
+        if not ret:
+            return
+        else:
+            xbmc.executebuiltin( 'ActivateWindow(busydialognocancel)' )
```

Comment:
with so many repetitions you can probably think of adding a contextlib entry in the future:

```python
with busydialog... :
    do_your_stuff()
```